### PR TITLE
MongoDB: support resizing flavor of a instance in replica-set.

### PR DESCRIPTION
**Pull request by caisan**
Enables user to resize the flavor of a instance that are the part of a
replica set cluster.

Currently this is only supported by MongoDB replica set. Before this PR,
try to resize a instance, which belongs to a cluster will raise
exception:ClusterInstanceOperationNotSupported.

Fixes: http://192.168.15.2/issues/11191

### CODE REVIEW

File: trove/common/strategies/cluster/experimental/mongodb/api.py [hunk 1]
```diff
@@ -321,6 +321,14 @@ def allow_backup(self):
         """
         return self.type == "replica-set"
 
+    def allow_resize_flavor(self):
+        """
+        Same as allow_backup(), Only resize the instance which is the part of
+        a replica-set cluster
+        :return: True is replica-set, False is not.
```
Comment (Farice4) [inline trove/common/strategies/cluster/experimental/mongodb/api.py]: return 描述信息有误?

File: trove/instance/models.py [hunk 2]
```diff
@@ -1019,7 +1020,14 @@ def resize_flavor(self, new_flavor_id):
                      "%(flavor_id)s."),
                  {'instance_id': self.id, 'flavor_id': new_flavor_id})
         if self.db_info.cluster_id is not None:
-            raise exception.ClusterInstanceOperationNotSupported()
+            """
+            Currently, only instance in MongoDB replica-set
+ +          cluster can execute backup. other cluster's instance cannot.
+            """
+            cluster = DBCluster.load(self.context,self.db_info.cluster_id)
```
Comment (Farice4) [inline trove/instance/models.py]: self.context,self.db_info.cluster_id 格式,中间空格?

File: trove/instance/models.py [hunk 3]
```diff
@@ -1019,7 +1020,14 @@ def resize_flavor(self, new_flavor_id):
                      "%(flavor_id)s."),
                  {'instance_id': self.id, 'flavor_id': new_flavor_id})
         if self.db_info.cluster_id is not None:
-            raise exception.ClusterInstanceOperationNotSupported()
+            """
+            Currently, only instance in MongoDB replica-set
+ +          cluster can execute backup. other cluster's instance cannot.
```
Comment (Farice4) [inline trove/instance/models.py]: 能否改为, "Support MongoDB replica-set  instance execute backup." ?

**Comment by Farice4** [comment]
commit 信息能否提炼, 如:
Fix xxxxxx

Add   xxxxx

Change  xxxxx

File: trove/instance/models.py (hunk 3 above)
Comment (caisan) [reply trove/instance/models.py]: typo. 应该把backup改成resize.

**Comment by caisan** [comment]
```from trove.cluster.models import Cluster as DBCluster```需要使用local import?

File: trove/common/strategies/cluster/experimental/mongodb/api.py [hunk 4]
```diff
@@ -321,6 +321,14 @@ def allow_backup(self):
         """
         return self.type == "replica-set"
 
+    def allow_resize_flavor(self):
```
Comment (zhaochao) [inline trove/common/strategies/cluster/experimental/mongodb/api.py]: 只是 resize_flavor ，resize_volume 不支持？

File: trove/instance/models.py [hunk 5]
```diff
@@ -50,6 +50,7 @@
 from trove.module import views as module_views
 from trove.quota.quota import run_with_quotas
 from trove.taskmanager import api as task_api
+from trove.cluster.models import Cluster as DBCluster
```
Comment (zhaochao) [inline trove/instance/models.py]: 是，要用 local import ，这个文件有一个现成的例子，在 L625 。

File: trove/instance/models.py [hunk 6]
```diff
@@ -1019,7 +1020,14 @@ def resize_flavor(self, new_flavor_id):
                      "%(flavor_id)s."),
                  {'instance_id': self.id, 'flavor_id': new_flavor_id})
         if self.db_info.cluster_id is not None:
-            raise exception.ClusterInstanceOperationNotSupported()
+            """
+            Currently, only instance in MongoDB replica-set
+ +          cluster can execute resize. other cluster's instance cannot.
+            """
+            cluster = DBCluster.load(self.context,self.db_info.cluster_id)
```
Comment (zhaochao) [inline trove/instance/models.py]: [bad-whitespace] Exactly one space required after comma [python/pylint]

**Review by zhaochao — CHANGES_REQUESTED** [review]
commit message 有不少地方要改一下：

> MongoDB: resize flavor instance in replica set
> 
> Enable user resize the flavor of instance that are the part of
> a cluster. Only enabled for MongoDB replica set. Before this, try to
> backup a instance, which belongs to a cluster will raise
> ception:ClusterInstanceOperationNotSupported.
> 
> Fix: http://192.168.15.2/issues/11191

第一行：应该大概是： MongoDB: support resizing flavor of a instance in replica set ；

Enable user resize -> Enables use to resize

the flavor of instance -> the flavor of a instance

a part of cluster -> a part of replica set cluster

Only enabled for MongoDB replica set. -> Currently this is only supported by MongoDB replica set.

ception: -> exception:

Fix: -> Fixes:

File: trove/common/strategies/cluster/experimental/mongodb/api.py (hunk 4 above)
Comment (caisan) [reply trove/common/strategies/cluster/experimental/mongodb/api.py]: @zhaochao 只是flavor.

File: trove/instance/models.py [hunk 7]
```diff
@@ -51,6 +51,7 @@
 from trove.quota.quota import run_with_quotas
 from trove.taskmanager import api as task_api
 
+
```
Comment (zhaochao) [inline trove/instance/models.py]: 无关改动？

**Review by zhaochao — CHANGES_REQUESTED** [review]
commit message 里 Fix: 也改成 Fixes: ，这样比较统一。

看了下 resize_flavor 的实现，目前看不出来会直接影响 replica set 的原因，改完剩下几个小问题可以先合并。

**Pull request merged**